src/dma/mod.rs

```diff
@@ -465,115 +448,128 @@ where
+    ///   returned by the closure does not match the current transfer size.
+    /// * The DMA may run into the poison address and throw an error if
+    ///   any of following conditions happen:
+    ///   * `SmallBuffer` error.
+    ///   * The closure `f` takes too long to return and a buffer overrun occurs.
+    pub fn next_transfer_with<F, T>(&mut self, func: F) -> Result<T, DMAError>
+    where
+        F: FnOnce(BUF, CurrentBuffer) -> (BUF, T),
+    {
+        let (double, inactive) = match STREAM::get_inactive_buffer() {
+            None => {
+                // Single buffer mode
+                self.stream.disable();
+                (false, CurrentBuffer::Buffer0)
             }
-
-            if STREAM::current_buffer() == CurrentBuffer::DoubleBuffer {
-                unsafe {
-                    self.stream.set_memory_address(new_buf_ptr as u32);
-                }
-                // Check if an overrun occurred, the buffer address won't be
-                // updated in that case
-                if self.stream.get_memory_address() != new_buf_ptr as u32 {
-                    self.stream.clear_transfer_complete_interrupt();
-                    return Err(DMAError::Overrun(new_buf));
+            Some(inactive) => {
+                // Double buffer mode
+                if !STREAM::get_transfer_complete_flag() {
+                    // DMA has not released a buffer
+                    return Err(DMAError::NotReady);
                 }
-
-                // "Subsequent reads and writes cannot be moved ahead of
-                // preceding reads"
-                compiler_fence(Ordering::Acquire);
-
-                let old_buf = self.buf.replace(new_buf);
-
-                // We always have a buffer, so unwrap can't fail
-                return Ok((old_buf.unwrap(), CurrentBuffer::FirstBuffer));
-            } else {
+                // Poison the peripheral's inactive memory address to get a memory
+                // error instead of potentially silent corruption if the DMA peripheral
+                // wins the race to the inactive buffer.
+                // NOTE(safety): Memory corruption will still occur in the "incactive"
+                // buffer if an DMA overrun occurs between reading the CT bit and poisoning
+                // the address.
                 unsafe {
-                    self.stream
-                        .set_memory_double_buffer_address(new_buf_ptr as u32);
-                }
-                // Check if an overrun occurred, the buffer address won't be
-                // updated in that case
-                if self.stream.get_memory_double_buffer_address()
-                    != new_buf_ptr as u32
-                {
-                    self.stream.clear_transfer_complete_interrupt();
-                    return Err(DMAError::Overrun(new_buf));
+                    self.stream.set_memory_address(inactive, 0xffff_ffffusize);
```

Comment:
It occurs to me that there's still a race condition here. If the stream swaps over directly after we check the active stream and before we update the address, the address write will fail (because the inactive buffer has become active).

A potential fix for this is to treat it like reading a two register timer - that is, we detect the inactive buffer and then attempt to "acquire" it by overwriting the buffer address. Then, we can check to see if the poisoned address sticks (write will fail if the buffer becomes active). We repeat that process until the poisoned address is accepted. Once we've managed to acquire one of the two buffers (by poisoning the associated address), we can continue with execution with a guarantee that there was no race condition.

I think it's dangerous to leave a race condition like this present in the HAL code otherwise.

Reply:
It's already commented and addressed. The poison write will lead to TEIF and disable the stream. There is no (silent, leading to corruption of an active buffer) race.
Even the case where DMA does several buffer swaps in the time we get to writing the poison (a few instructions) is the same as the case where we just handle the TCIF late.

Reply:
There's technically still a case here where the DMA interrupt occurs some multiple of 2 times between getting the inactive buffer and setting the memory address. This might occur if an ISR happens between these two functions.

The result would be that a buffer would be overwritten (potentially multiple times) before the swap occurs.

Reply:
That's what I just said in my the last sentence. But it's not a separate problem from the general case (which should be now described and documented in abundant detail).
Or am I misunderstanding you?

Reply:
Yeah, I agree with you here. Closing this as resolved - I don't think we can justifiably worry about that at this level.